### DISCUSSION TITLE
1D ODE with non-constant coefficients.

### DISCUSSION THREAD

**Discussion by ethrelfall** [firedrakeproject/firedrake#2686] [Firedrake support] [answered]
Dear community,

I've been trying to set up a very simple example of a 1D ODE in weak form.  The equation contains the radial part of the scalar Laplacian in 2D and so there are metric scale factors giving non-constant coefficients.

I have implemented something that I think is plausible but it doesn't give the correct solution.  The (very short) script and the problem are explained at https://github.com/ethrelfall/further-Firedrake.

Please, is someone able to say how to do this properly?

(Search for matching discussions returned nowt.)

Thanks!

**Comment by colinjcotter** [comment]
Code looks ok at a glance. Are you comparing with the exact solution applied to w, not the coordinate, by any chance?

________________________________________
From: ethrelfall ***@***.***>
Sent: 13 December 2022 17:19
To: firedrakeproject/firedrake
Cc: Subscribed
Subject: [firedrakeproject/firedrake] 1D ODE with non-constant coefficients. (Discussion #2686)

Dear community,

I've been trying to set up a very simple example of a 1D ODE in weak form. The equation contains the radial part of the scalar Laplacian in 2D and so there are metric scale factors giving non-constant coefficients.

I have implemented something that I think is plausible but it doesn't give the correct solution. The (very short) script and the problem are explained at https://github.com/ethrelfall/further-Firedrake.

Please, is someone able to say how to do this properly?

(Search for matching discussions returned nowt.)

Thanks!

**Comment by colinjcotter** [comment]
Oh, I see the problem. Your RHS is v*w*dx but it should be q*v*dx where q is set to -1.

________________________________________

**Reply by ethrelfall** [reply]
Thanks so much for taking a look.  But why should I NOT have \int r v dx on the RHS (wh. w=r in the script), as in the formula in the readme - ?

**Comment by ksagiyam** [comment, marked as answer]
Your analytical solution is not correct; if I add a correction term to your rhs as in `L = ((w+1.5/w**2)*v)*dx`, what you wrote down becomes the exact solution.

**Reply by ethrelfall** [reply]
Ah you're right, I see I used 1/r in the analytic solution when it should have been ln r ... oops!  Thanks v. much.

**Comment by colinjcotter** [comment]
oh ah you multiplied through by r! sorry!

On 13 Dec 2022, at 18:54, ethrelfall ***@***.***> wrote:

﻿

Thanks so much for taking a look. But why should I NOT have \int r v dx on the RHS (wh. w=r in the script), as in the formula in the readme - ?

**Reply by ethrelfall** [reply]
Yes then I made schoolboy error :(
